src/libsystemd-network/sd-dhcp-client.c

```diff
@@ -1675,49 +1661,17 @@ static int client_set_lease_timeouts(sd_dhcp_client *client) {
-                        client->t1_time = client_compute_timeout(client, client->lease->t1, 1);
-                } else {
-                        /* discard both */
-                        client->t2_time = client_compute_timeout(client, client->lease->lifetime, 7.0 / 8.0);
-                        client->lease->t2 = (client->lease->lifetime * 7) / 8;
-                        client->t1_time = client_compute_timeout(client, client->lease->lifetime, 0.5);
-                        client->lease->t1 = client->lease->lifetime / 2;
-                }
-        } else if (client->lease->t2 > 0 && client->lease->t2 < client->lease->lifetime) {
-                /* only T2 is given, and it is valid */
-                client->t2_time = client_compute_timeout(client, client->lease->t2, 1);
-                client->t1_time = client_compute_timeout(client, client->lease->lifetime, 0.5);
-                client->lease->t1 = client->lease->lifetime / 2;
-                if (client->t2_time <= client->t1_time) {
-                        /* the computed T1 would be invalid, so discard T2 */
-                        client->t2_time = client_compute_timeout(client, client->lease->lifetime, 7.0 / 8.0);
-                        client->lease->t2 = (client->lease->lifetime * 7) / 8;
-                }
-        } else if (client->lease->t1 > 0 && client->lease->t1 < client->lease->lifetime) {
-                /* only T1 is given, and it is valid */
-                client->t1_time = client_compute_timeout(client, client->lease->t1, 1);
-                client->t2_time = client_compute_timeout(client, client->lease->lifetime, 7.0 / 8.0);
-                client->lease->t2 = (client->lease->lifetime * 7) / 8;
-                if (client->t2_time <= client->t1_time) {
-                        /* the computed T2 would be invalid, so discard T1 */
-                        client->t2_time = client_compute_timeout(client, client->lease->lifetime, 0.5);
-                        client->lease->t2 = client->lease->lifetime / 2;
-                }
-        } else {
-                /* fall back to the default timeouts */
-                client->t1_time = client_compute_timeout(client, client->lease->lifetime, 0.5);
-                client->lease->t1 = client->lease->lifetime / 2;
-                client->t2_time = client_compute_timeout(client, client->lease->lifetime, 7.0 / 8.0);
-                client->lease->t2 = (client->lease->lifetime * 7) / 8;
-        }
+        /* set default t1/t2 if unset or invalid */
+        if (client->lease->t2 == 0 || client->lease->t2 >= client->lease->lifetime)
+                client->lease->t2 = RFC2131_T2_DEFAULT(client->lease->lifetime);
+        if (client->lease->t1 == 0 || client->lease->t1 >= client->lease->t2)
+                client->lease->t1 = RFC2131_T1_DEFAULT(client->lease->lifetime);
+        if (client->lease->t1 >= client->lease->t2)
+                client->lease->t2 = RFC2131_T2_DEFAULT(client->lease->lifetime);
```

Comment:
Still there exist possibilities that `t2 <= t1` or `lifetime <= t1`...

Reply:
I may be missing something, but here is my thinking (and this assumes that T1_DEFAULT() < T2_DEFAULT() < lifetime):

```
        if (client->lease->t2 == 0 || client->lease->t2 >= client->lease->lifetime)                                                                                                                                                                                           
                client->lease->t2 = T2_DEFAULT(client->lease->lifetime);                                                                                                                                                                                                      

```
this guarantees that 0 < t2 < lifetime

```
        if (client->lease->t1 == 0 || client->lease->t1 >= client->lease->t2)                                                                                                                                                                                                 
                client->lease->t1 = T1_DEFAULT(client->lease->lifetime);                                                                                                                                                                                                      

```

since we know t2 < lifetime, this guarantees that 0 < t1 < lifetime

```
        if (client->lease->t1 >= client->lease->t2)                                                                                                                                                                                                                           
                client->lease->t2 = T2_DEFAULT(client->lease->lifetime);                                                                                                                                                                                                      

```

if this is true, and t1 >= t2, then t1 *must* be T1_DEFAULT, since the previous check could not have evaluated to false. in that case, assigning t2 to T2_DEFAULT is guaranteed to make t1 < t2.